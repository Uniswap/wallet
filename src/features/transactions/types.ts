import { TradeType } from '@uniswap/sdk-core'
import { providers } from 'ethers'
import { ChainId, ChainIdTo } from 'src/constants/chains'
import { TransactionListQuery } from 'src/data/__generated__/types-and-hooks'
import { AssetType } from 'src/entities/assets'
import { MoonpayCurrency } from 'src/features/fiatOnRamp/types'
import { DappInfo } from 'src/features/walletConnect/types'

export type ChainIdToTxIdToDetails = ChainIdTo<{ [txId: string]: TransactionDetails }>

// Basic identifying info for a transaction
export interface TransactionId {
  chainId: ChainId
  id: string
}

export type TransactionListQueryResponse = NonNullable<
  NonNullable<NonNullable<TransactionListQuery['portfolios']>[0]>['assetActivities']
>[0]

export interface TransactionDetails extends TransactionId {
  from: Address

  // Specific info for the tx type
  typeInfo: TransactionTypeInfo

  // Info for status tracking
  status: TransactionStatus
  addedTime: number
  // Note: hash is mandatory for now but may be made optional if
  // we start tracking txs before they're actually sent
  hash: string

  // Info for submitting the tx
  options: TransactionOptions

  receipt?: TransactionReceipt

  isFlashbots?: boolean

  // cancelRequest is the txRequest object to be submitted
  // in attempt to cancel the current transaction
  // it should contain all the appropriate gas details in order
  // to get submitted first
  cancelRequest?: providers.TransactionRequest
}

export enum TransactionStatus {
  Cancelled = 'cancelled',
  Cancelling = 'cancelling',
  FailedCancel = 'failedCancel',
  Success = 'confirmed',
  Failed = 'failed',
  Pending = 'pending',
  Replacing = 'replacing',
  Unknown = 'unknown',
  // May want more granular options here later like InMemPool
}

// Transaction confirmed on chain
export type FinalizedTransactionStatus =
  | TransactionStatus.Success
  | TransactionStatus.Failed
  | TransactionStatus.Cancelled
  | TransactionStatus.FailedCancel

export interface FinalizedTransactionDetails extends TransactionDetails {
  status: FinalizedTransactionStatus
}

export interface TransactionOptions {
  request: providers.TransactionRequest
  timeoutMs?: number
}

export interface TransactionReceipt {
  transactionIndex: number
  blockHash: string
  blockNumber: number
  confirmedTime: number
  confirmations: number
}

export interface NFTSummaryInfo {
  tokenId: string
  name: string
  collectionName: string
  imageURL: string
}

export enum NFTTradeType {
  BUY = 'buy',
  SELL = 'sell',
}

/**
 * Be careful adding to this enum, always assign a unique value (typescript will not prevent duplicate values).
 * These values are persisted in state and if you change the value it will cause errors
 */
export enum TransactionType {
  // Token Specific
  Approve = 'approve',
  Swap = 'swap',
  Wrap = 'wrap',

  // NFT specific
  NFTApprove = 'nft-approve',
  NFTTrade = 'nft-trade',
  NFTMint = 'nft-mint',

  // All asset types
  Send = 'send',
  Receive = 'receive',

  // Fiat onramp
  FiatPurchase = 'fiat-purchase',

  // General
  WCConfirm = 'wc-confirm',
  Unknown = 'unknown',
}

export interface BaseTransactionInfo {
  type: TransactionType
  transactedUSDValue?: number
}

export interface ApproveTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.Approve
  tokenAddress: string
  spender: string
  approvalAmount?: string
}

interface BaseSwapTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.Swap
  tradeType: TradeType
  inputCurrencyId: string
  outputCurrencyId: string
}

export interface ExactInputSwapTransactionInfo extends BaseSwapTransactionInfo {
  tradeType: TradeType.EXACT_INPUT
  inputCurrencyAmountRaw: string
  expectedOutputCurrencyAmountRaw: string
  minimumOutputCurrencyAmountRaw: string
}

export interface ExactOutputSwapTransactionInfo extends BaseSwapTransactionInfo {
  tradeType: TradeType.EXACT_OUTPUT
  outputCurrencyAmountRaw: string
  expectedInputCurrencyAmountRaw: string
  maximumInputCurrencyAmountRaw: string
}

export interface WrapTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.Wrap
  unwrapped: boolean
  currencyAmountRaw: string
}

export interface SendTokenTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.Send
  assetType: AssetType
  recipient: string
  tokenAddress: string
  currencyAmountRaw?: string
  tokenId?: string // optional. NFT token id
  nftSummaryInfo?: NFTSummaryInfo // optional. NFT metadata
}

export interface ReceiveTokenTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.Receive
  assetType: AssetType
  currencyAmountRaw?: string
  sender: string
  tokenAddress: string
  tokenId?: string // optional. NFT token id
  nftSummaryInfo?: NFTSummaryInfo
  isSpam?: boolean
}

export interface FiatPurchaseTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.FiatPurchase
  explorerUrl?: string
  // code will be used for formatting amounts
  inputCurrency?: Pick<MoonpayCurrency, 'type' | 'code'>
  inputCurrencyAmount?: number
  // metadata will be used to get the output currency
  outputCurrency?: Required<Pick<MoonpayCurrency, 'type' | 'metadata'>>
  outputCurrencyAmount?: number
  syncedWithBackend: boolean
}

export interface NFTMintTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.NFTMint
  nftSummaryInfo: NFTSummaryInfo
  purchaseCurrencyId?: string
  purchaseCurrencyAmountRaw?: string
}

export interface NFTTradeTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.NFTTrade
  nftSummaryInfo: NFTSummaryInfo
  purchaseCurrencyId: string
  purchaseCurrencyAmountRaw: string
  tradeType: NFTTradeType
}

export interface NFTApproveTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.NFTApprove
  nftSummaryInfo: NFTSummaryInfo
  spender: string
}

export interface WCConfirmInfo extends BaseTransactionInfo {
  type: TransactionType.WCConfirm
  dapp: DappInfo
}

export interface UnknownTransactionInfo extends BaseTransactionInfo {
  type: TransactionType.Unknown
  tokenAddress?: string
}

export type TransactionTypeInfo =
  | ApproveTransactionInfo
  | FiatPurchaseTransactionInfo
  | ExactOutputSwapTransactionInfo
  | ExactInputSwapTransactionInfo
  | WrapTransactionInfo
  | SendTokenTransactionInfo
  | ReceiveTokenTransactionInfo
  | NFTTradeTransactionInfo
  | NFTApproveTransactionInfo
  | NFTMintTransactionInfo
  | WCConfirmInfo
  | UnknownTransactionInfo
