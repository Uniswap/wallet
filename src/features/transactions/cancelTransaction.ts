import { CallEffect, SelectEffect } from 'redux-saga/effects'
import { appSelect } from 'src/app/hooks'
import { attemptReplaceTransaction } from 'src/features/transactions/replaceTransaction'
import { makeSelectTransaction } from 'src/features/transactions/selectors'
import { TransactionDetails } from 'src/features/transactions/types'
import { call } from 'typed-redux-saga'

// Note, transaction cancellation on Ethereum is inherently flaky
// The best we can do is replace the transaction and hope the original isn't mined first
// Inspiration: https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/transactions/index.js#L744
export function* attemptCancelTransaction(
  transaction: TransactionDetails
): Generator<SelectEffect | CallEffect<void>, void, unknown> {
  const { from, chainId, id } = transaction
  // get updated TransactionDetails object to query for cancelRequest
  // TODO(MOB-3968): Add more specific type definition here
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const tx: any = yield* appSelect(makeSelectTransaction(from, chainId, id))
  if (!tx?.cancelRequest) {
    throw new Error('attempted to cancel a transaction without cancelRequest set')
  }
  yield* call(attemptReplaceTransaction, transaction, tx.cancelRequest, true)
}
