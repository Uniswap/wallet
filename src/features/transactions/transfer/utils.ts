import { AnyAction } from '@reduxjs/toolkit'
import {
  selectRecipient,
  toggleShowRecipientSelector,
} from 'src/features/transactions/transactionState/transactionState'

export const createOnToggleShowRecipientSelector =
  (dispatch: React.Dispatch<AnyAction>) => (): void =>
    dispatch(toggleShowRecipientSelector)

export const createOnSelectRecipient = (dispatch: React.Dispatch<AnyAction>) => {
  return (recipient: Address): void => {
    createOnToggleShowRecipientSelector(dispatch)()
    dispatch(selectRecipient({ recipient }))
  }
}
