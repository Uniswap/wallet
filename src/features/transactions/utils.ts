import { CurrencyAmount, NativeCurrency, TradeType } from '@uniswap/sdk-core'
import { BigNumber, providers } from 'ethers'
import { ChainId } from 'src/constants/chains'
import {
  ExactInputSwapTransactionInfo,
  ExactOutputSwapTransactionInfo,
} from 'src/features/transactions/types'
import { v4 as uuid } from 'uuid'

export function getSerializableTransactionRequest(
  request: providers.TransactionRequest,
  chainId?: ChainId
): providers.TransactionRequest {
  // prettier-ignore
  const { to, from, nonce, gasLimit, gasPrice, data, value, maxPriorityFeePerGas, maxFeePerGas, type } = request
  // Manually restructure the txParams to ensure values going into store are serializable
  return {
    chainId,
    type,
    to,
    from,
    nonce: nonce ? BigNumber.from(nonce).toString() : undefined,
    gasLimit: gasLimit?.toString(),
    gasPrice: gasPrice?.toString(),
    data: data?.toString(),
    value: value?.toString(),
    maxPriorityFeePerGas: maxPriorityFeePerGas?.toString(),
    maxFeePerGas: maxFeePerGas?.toString(),
  }
}

function getNativeCurrencyTotalSpend(
  value?: CurrencyAmount<NativeCurrency>,
  gasFee?: string,
  nativeCurrency?: NativeCurrency
): NullUndefined<CurrencyAmount<NativeCurrency>> {
  if (!gasFee || !nativeCurrency) return value

  const gasFeeAmount = CurrencyAmount.fromRawAmount(nativeCurrency, gasFee)
  return value ? gasFeeAmount.add(value) : gasFeeAmount
}

export function hasSufficientFundsIncludingGas(params: {
  transactionAmount?: CurrencyAmount<NativeCurrency>
  gasFee?: string
  nativeCurrencyBalance?: CurrencyAmount<NativeCurrency>
}): boolean {
  const { transactionAmount, gasFee, nativeCurrencyBalance } = params
  const totalSpend = getNativeCurrencyTotalSpend(
    transactionAmount,
    gasFee,
    nativeCurrencyBalance?.currency
  )
  return !totalSpend || !nativeCurrencyBalance?.lessThan(totalSpend)
}

export function createTransactionId(): string {
  return uuid()
}

export function getInputAmountFromTrade(
  typeInfo: ExactInputSwapTransactionInfo | ExactOutputSwapTransactionInfo
): string {
  return typeInfo.tradeType === TradeType.EXACT_INPUT
    ? typeInfo.inputCurrencyAmountRaw
    : typeInfo.expectedInputCurrencyAmountRaw
}

export function getOutputAmountFromTrade(
  typeInfo: ExactInputSwapTransactionInfo | ExactOutputSwapTransactionInfo
): string {
  return typeInfo.tradeType === TradeType.EXACT_OUTPUT
    ? typeInfo.outputCurrencyAmountRaw
    : typeInfo.expectedOutputCurrencyAmountRaw
}

export const ANIMATE_SPRING_CONFIG = {
  stiffness: 90,
  damping: 15,
  mass: 0.8,
}
