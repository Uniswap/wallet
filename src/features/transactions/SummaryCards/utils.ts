import { TFunction } from 'i18next'
import { TransactionStatus } from 'src/features/transactions/types'

export function getTransactionTitle(
  status: TransactionStatus,
  presentText: string, // present tense form of the title
  pastText: string | undefined, // past tense form of the title
  t: TFunction
): string {
  const prefixFail = status === TransactionStatus.Failed ? t('Failed') : ''

  // For items with pending or alert banner UI (see AlertBanner.tsx), use present tense.
  const isPresentTense =
    status === TransactionStatus.Pending ||
    status === TransactionStatus.Cancelling ||
    status === TransactionStatus.Cancelled ||
    !!prefixFail

  const baseText = isPresentTense || !pastText ? presentText : pastText

  // Lowercase title if prefix
  const title =
    prefixFail + (prefixFail ? ' ' : '') + (prefixFail ? baseText.toLocaleLowerCase() : baseText)

  return title
}
