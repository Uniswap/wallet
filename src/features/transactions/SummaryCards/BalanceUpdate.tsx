import { Currency } from '@uniswap/sdk-core'
import React, { useMemo } from 'react'
import { useSpotPrice } from 'src/features/dataApi/spotPricesQuery'
import { createBalanceUpdate } from 'src/features/notifications/utils'
import { AssetUpdateLayout } from 'src/features/transactions/SummaryCards/TransactionSummaryLayout'
import { NFTTradeType, TransactionStatus, TransactionType } from 'src/features/transactions/types'
import { currencyId } from 'src/utils/currencyId'

interface BalanceUpdateProps {
  currency: Currency
  amountRaw: string
  transactionType: TransactionType
  transactionStatus: TransactionStatus
  nftTradeType?: NFTTradeType
  transactedUSDValue?: number // Overrides API request for spot price if USD value already known
}

export default function BalanceUpdate({
  currency,
  amountRaw,
  transactionType,
  transactionStatus,
  nftTradeType,
  transactedUSDValue,
}: BalanceUpdateProps): JSX.Element | null {
  // Skip fetching spot price if known USD transacted value
  const skipQuery = !!transactedUSDValue
  const _currencyId = currencyId(currency)
  const { data: spotPrice, loading } = useSpotPrice(_currencyId, skipQuery)

  const balanceUpdate = useMemo(() => {
    if (!amountRaw || loading) {
      return undefined
    }
    return createBalanceUpdate({
      transactionType,
      transactionStatus,
      currency,
      currencyAmountRaw: amountRaw,
      spotPrice,
      nftTradeType,
      transactedUSDValue,
    })
  }, [
    amountRaw,
    currency,
    nftTradeType,
    spotPrice,
    transactedUSDValue,
    transactionStatus,
    transactionType,
    loading,
  ])

  if (loading) return null

  return (
    <AssetUpdateLayout
      caption={balanceUpdate?.usdValueChange}
      title={balanceUpdate?.assetValueChange}
    />
  )
}
