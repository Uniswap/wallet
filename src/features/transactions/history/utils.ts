import dayjs from 'dayjs'
import { BigNumber } from 'ethers'
import { parseUnits } from 'ethers/lib/utils'
import { ChainId } from 'src/constants/chains'
import {
  Amount,
  Chain,
  Currency,
  TokenStandard,
  TransactionListQuery,
} from 'src/data/__generated__/types-and-hooks'
import { NativeCurrency } from 'src/features/tokens/NativeCurrency'
import extractTransactionDetails from 'src/features/transactions/history/conversion/extractTransactionDetails'
import {
  TransactionDetails,
  TransactionStatus,
  TransactionType,
} from 'src/features/transactions/types'
import { fromGraphQLChain } from 'src/utils/chainId'
import { getNativeCurrencyAddressForChain } from 'src/utils/currencyId'

export interface AllFormattedTransactions {
  combinedTransactionList: TransactionDetails[]
  todayTransactionList: TransactionDetails[]
  monthTransactionList: TransactionDetails[]
  // Maps year <-> TransactionSummaryInfo[] for all months before current month
  priorByMonthTransactionList: Record<string, TransactionDetails[]>
  pending: TransactionDetails[]
}

export function formatTransactionsByDate(
  transactions: TransactionDetails[]
): AllFormattedTransactions {
  // timestamp in ms for start of time periods
  const msTimestampCutoffDay = dayjs().startOf('day').unix() * 1000
  const msTimestampCutoffMonth = dayjs().startOf('month').unix() * 1000
  const msTimestampCutoffYear = dayjs().startOf('year').unix() * 1000

  // Segement by time periods.
  const [pending, todayTransactionList, monthTransactionList, beforeCurrentMonth] =
    transactions.reduce<
      [TransactionDetails[], TransactionDetails[], TransactionDetails[], TransactionDetails[]]
    >(
      (accum, item) => {
        if (
          // Want all incomplete transactions
          item.status === TransactionStatus.Pending ||
          item.status === TransactionStatus.Cancelling ||
          item.status === TransactionStatus.Replacing
        ) {
          accum[0].push(item)
        } else if (item.addedTime > msTimestampCutoffDay) {
          accum[1].push(item)
        } else if (item.addedTime > msTimestampCutoffMonth) {
          accum[2].push(item)
        } else {
          accum[3].push(item)
        }
        return accum
      },
      [[], [], [], []]
    )

  // sort pending txns based on nonce, highest nonce first for reverse chronological order
  const pendingSorted = pending.sort((a, b) => {
    const nonceA = a.options?.request?.nonce
    const nonceB = b.options?.request?.nonce
    return nonceA && nonceB ? (nonceA < nonceB ? 1 : -1) : 1
  })

  // For all transactions before current month, group by month
  const priorByMonthTransactionList = beforeCurrentMonth.reduce(
    (accum: Record<string, TransactionDetails[]>, item) => {
      const isPreviousYear = item.addedTime < msTimestampCutoffYear
      const key = dayjs(item.addedTime)
        // If in a previous year, append year to key string, else just use month
        // This key is used as the section title in TransactionList
        .format(isPreviousYear ? 'MMMM YYYY' : 'MMMM')
        .toString()
      const currentMonthList = accum[key] ?? []
      currentMonthList.push(item)
      accum[key] = currentMonthList
      return accum
    },
    {}
  )

  return {
    combinedTransactionList: transactions,
    pending: pendingSorted,
    todayTransactionList,
    monthTransactionList,
    priorByMonthTransactionList,
  }
}

/**
 * Transforms api txn data to formatted TransactionDetails array
 * @param data Transaction history data response
 */
export function parseDataResponseToTransactionDetails(
  data: TransactionListQuery,
  hideSpamTokens?: boolean
): TransactionDetails[] {
  if (data.portfolios?.[0]?.assetActivities) {
    return data.portfolios[0].assetActivities.reduce((accum: TransactionDetails[], t) => {
      const parsed = extractTransactionDetails(t)

      // Filter out spam if desired, currently only for receive transactions
      const isSpam = parsed?.typeInfo.type === TransactionType.Receive && parsed.typeInfo.isSpam

      if (parsed && !(hideSpamTokens && isSpam)) {
        accum.push(parsed)
      }

      return accum
    }, [])
  }
  return []
}

/**
 * Constructs a CurrencyAmount based on asset details and quantity. Checks if token is native
 * or ERC20 to determine decimal amount.
 * @param tokenStandard token standard type from api query
 * @param quantity // formatted amount of asset transfered
 * @param decimals // decimals ((optional) if native token)
 * @returns
 */
export function deriveCurrencyAmountFromAssetResponse(
  tokenStandard: TokenStandard,
  quantity: string,
  decimals: NullUndefined<number>
): string {
  return parseUnits(
    quantity,
    tokenStandard === TokenStandard.Native
      ? BigNumber.from(NativeCurrency.onChain(ChainId.Mainnet).decimals)
      : decimals
      ? BigNumber.from(decimals)
      : undefined
  ).toString()
}

/**
 * Parses an asset from API and returns either the token address or native currency address
 * for the involved asset.
 * @returns Token address, custom native address or null
 */
export function getAddressFromAsset({
  tokenStandard,
  chain,
  address,
}: {
  tokenStandard: TokenStandard
  chain: Chain | undefined
  address: NullUndefined<string>
}): NullUndefined<string> {
  const supportedChainId = fromGraphQLChain(chain)
  if (!supportedChainId) {
    return null
  }
  if (tokenStandard === TokenStandard.Native) {
    return getNativeCurrencyAddressForChain(supportedChainId)
  }
  return address
}

/**
 *
 * @param transactedValue Transacted value amount from TokenTransfer API response
 * @returns parsed USD value as a number if currency is of type USD
 */
export function parseUSDValueFromAssetChange(
  transactedValue: NullUndefined<Partial<Amount>>
): number | undefined {
  return transactedValue?.currency === Currency.Usd ? transactedValue.value ?? undefined : undefined
}
