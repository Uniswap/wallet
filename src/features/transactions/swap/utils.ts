import { Currency, TradeType } from '@uniswap/sdk-core'
import { BigNumber, BigNumberish } from 'ethers'
import { TFunction } from 'i18next'
import { ChainId } from 'src/constants/chains'
import { WRAPPED_NATIVE_CURRENCY } from 'src/constants/tokens'
import { AssetType } from 'src/entities/assets'
import { DEFAULT_SLIPPAGE_TOLERANCE_PERCENT } from 'src/features/transactions/swap/hooks'
import { Trade } from 'src/features/transactions/swap/useTrade'
import { WrapType } from 'src/features/transactions/swap/wrapSaga'
import {
  ExactInputSwapTransactionInfo,
  ExactOutputSwapTransactionInfo,
  TransactionType,
} from 'src/features/transactions/types'
import { areAddressesEqual } from 'src/utils/addresses'
import {
  CurrencyId,
  currencyId,
  currencyIdToAddress,
  currencyIdToChain,
} from 'src/utils/currencyId'
import { formatPrice, NumberType } from 'src/utils/format'
import { CurrencyField, TransactionState } from '../transactionState/transactionState'

export function serializeQueryParams(
  params: Record<string, Parameters<typeof encodeURIComponent>[0]>
): string {
  const queryString = []
  for (const [param, value] of Object.entries(params)) {
    queryString.push(`${encodeURIComponent(param)}=${encodeURIComponent(value)}`)
  }
  return queryString.join('&')
}

export function getWrapType(
  inputCurrency: Currency | null | undefined,
  outputCurrency: Currency | null | undefined
): WrapType {
  if (!inputCurrency || !outputCurrency || inputCurrency.chainId !== outputCurrency.chainId) {
    return WrapType.NotApplicable
  }

  const weth = WRAPPED_NATIVE_CURRENCY[inputCurrency.chainId as ChainId]

  if (inputCurrency.isNative && outputCurrency.equals(weth)) {
    return WrapType.Wrap
  } else if (outputCurrency.isNative && inputCurrency.equals(weth)) {
    return WrapType.Unwrap
  }

  return WrapType.NotApplicable
}

export function isWrapAction(wrapType: WrapType): wrapType is WrapType.Unwrap | WrapType.Wrap {
  return wrapType === WrapType.Unwrap || wrapType === WrapType.Wrap
}

export function tradeToTransactionInfo(
  trade: Trade
): ExactInputSwapTransactionInfo | ExactOutputSwapTransactionInfo {
  return trade.tradeType === TradeType.EXACT_INPUT
    ? {
        type: TransactionType.Swap,
        inputCurrencyId: currencyId(trade.inputAmount.currency),
        outputCurrencyId: currencyId(trade.outputAmount.currency),
        tradeType: TradeType.EXACT_INPUT,
        inputCurrencyAmountRaw: trade.inputAmount.quotient.toString(),
        expectedOutputCurrencyAmountRaw: trade.outputAmount.quotient.toString(),
        minimumOutputCurrencyAmountRaw: trade
          .minimumAmountOut(DEFAULT_SLIPPAGE_TOLERANCE_PERCENT)
          .quotient.toString(),
      }
    : {
        type: TransactionType.Swap,
        inputCurrencyId: currencyId(trade.inputAmount.currency),
        outputCurrencyId: currencyId(trade.outputAmount.currency),
        tradeType: TradeType.EXACT_OUTPUT,
        outputCurrencyAmountRaw: trade.outputAmount.quotient.toString(),
        expectedInputCurrencyAmountRaw: trade.inputAmount.quotient.toString(),
        maximumInputCurrencyAmountRaw: trade
          .maximumAmountIn(DEFAULT_SLIPPAGE_TOLERANCE_PERCENT)
          .quotient.toString(),
      }
}

export function requireAcceptNewTrade(
  oldTrade: NullUndefined<Trade>,
  newTrade: NullUndefined<Trade>
): boolean {
  return oldTrade?.quote?.methodParameters?.calldata !== newTrade?.quote?.methodParameters?.calldata
}

export const getRateToDisplay = (trade: Trade, showInverseRate: boolean): string => {
  const price = showInverseRate ? trade.executionPrice.invert() : trade.executionPrice
  const formattedPrice = formatPrice(price, NumberType.SwapPrice)
  const { quoteCurrency, baseCurrency } = trade.executionPrice
  const rate = `1 ${quoteCurrency.symbol} = ${formattedPrice} ${baseCurrency.symbol}`
  const inverseRate = `1 ${baseCurrency.symbol} = ${formattedPrice} ${quoteCurrency.symbol}`
  return showInverseRate ? rate : inverseRate
}

export const formatAsHexString = (input?: BigNumberish): string | undefined =>
  input !== undefined ? BigNumber.from(input).toHexString() : input

export const getActionName = (t: TFunction, wrapType: WrapType): string => {
  switch (wrapType) {
    case WrapType.Unwrap:
      return t('Unwrap')
    case WrapType.Wrap:
      return t('Wrap')
    default:
      return t('Swap')
  }
}

export const getReviewActionName = (t: TFunction, wrapType: WrapType): string => {
  switch (wrapType) {
    case WrapType.Unwrap:
      return t('Review unwrap')
    case WrapType.Wrap:
      return t('Review wrap')
    default:
      return t('Review swap')
  }
}

export function sumGasFees(gasFee1?: string | undefined, gasFee2?: string): string | undefined {
  if (!gasFee1 || !gasFee2) return gasFee1 || gasFee2

  return BigNumber.from(gasFee1).add(gasFee2).toString()
}

export const clearStaleTrades = (
  trade: Trade,
  currencyIn: NullUndefined<Currency>,
  currencyOut: NullUndefined<Currency>
): Trade | null => {
  const currencyInAddress = currencyIn?.wrapped.address
  const currencyOutAddress = currencyOut?.wrapped.address

  const inputsMatch =
    !!currencyInAddress &&
    areAddressesEqual(currencyInAddress, trade?.inputAmount.currency.wrapped.address)
  const outputsMatch =
    !!currencyOutAddress &&
    areAddressesEqual(currencyOutAddress, trade?.outputAmount.currency.wrapped.address)

  // if the addresses entered by the user don't match what is being returned by the quote endpoint
  // then set `trade` to null
  return inputsMatch && outputsMatch ? trade : null
}

export const prepareSwapFormState = ({
  inputCurrencyId,
}: {
  inputCurrencyId?: CurrencyId
}): TransactionState | undefined => {
  return inputCurrencyId
    ? {
        exactCurrencyField: CurrencyField.INPUT,
        exactAmountToken: '',
        [CurrencyField.INPUT]: {
          address: currencyIdToAddress(inputCurrencyId),
          chainId: currencyIdToChain(inputCurrencyId) ?? ChainId.Mainnet,
          type: AssetType.Currency,
        },
        [CurrencyField.OUTPUT]: null,
      }
    : undefined
}
