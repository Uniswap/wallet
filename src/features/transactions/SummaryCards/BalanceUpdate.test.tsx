import { MockedResponse } from '@apollo/client/testing'
import React from 'react'
import { ChainId } from 'src/constants/chains'
import { SpotPricesDocument, SpotPricesQuery } from 'src/data/__generated__/types-and-hooks'
import { currencyIdToContractInput } from 'src/features/dataApi/utils'
import { NativeCurrency } from 'src/features/tokens/NativeCurrency'
import BalanceUpdate from 'src/features/transactions/SummaryCards/BalanceUpdate'
import { TransactionStatus, TransactionType } from 'src/features/transactions/types'
import { TokenProjects } from 'src/test/gqlFixtures'
import { render, screen } from 'src/test/test-utils'
import { currencyId } from 'src/utils/currencyId'

const ETH = NativeCurrency.onChain(ChainId.Mainnet)
const _currencyId = currencyId(ETH)

const mock: MockedResponse<SpotPricesQuery> = {
  request: {
    query: SpotPricesDocument,
    variables: {
      contracts: [currencyIdToContractInput(_currencyId)],
    },
  },
  result: {
    data: {
      tokenProjects: TokenProjects,
    },
  },
}

describe(BalanceUpdate, () => {
  it('renders without error after fetching data', async () => {
    const tree = render(
      <BalanceUpdate
        amountRaw="10000.00"
        currency={ETH}
        transactionStatus={TransactionStatus.Success}
        transactionType={TransactionType.Receive}
      />,
      { mocks: [mock] }
    )

    // loading
    expect(tree.toJSON()).toBeNull()

    // success
    expect(await screen.findByText('+ETH')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('skips fetching data when passing in the USD value', async () => {
    const tree = render(
      <BalanceUpdate
        amountRaw="10000.00"
        currency={ETH}
        transactedUSDValue={100}
        transactionStatus={TransactionStatus.Success}
        transactionType={TransactionType.Receive}
      />,
      { mocks: [mock] }
    )

    // no loading because the call is skipped
    expect(screen.getByText('$100.00')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
