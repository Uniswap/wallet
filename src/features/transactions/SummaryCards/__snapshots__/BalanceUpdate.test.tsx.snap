// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalanceUpdate renders without error after fetching data 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-end",
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#0D111C",
          "display": "flex",
          "fontFamily": "Inter-Medium",
          "fontSize": 16,
          "lineHeight": 24,
        },
      ]
    }
  >
    +ETH
  </Text>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "height": 0,
        },
      ]
    }
  />
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#505A78",
          "display": "flex",
          "fontFamily": "Inter-Regular",
          "fontSize": 14,
          "lineHeight": 20,
        },
      ]
    }
  >
    &lt;$0.00000001
  </Text>
</View>
`;

exports[`BalanceUpdate skips fetching data when passing in the USD value 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-end",
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#0D111C",
          "display": "flex",
          "fontFamily": "Inter-Medium",
          "fontSize": 16,
          "lineHeight": 24,
        },
      ]
    }
  >
    +ETH
  </Text>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "height": 0,
        },
      ]
    }
  />
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#505A78",
          "display": "flex",
          "fontFamily": "Inter-Regular",
          "fontSize": 14,
          "lineHeight": 20,
        },
      ]
    }
  >
    $100.00
  </Text>
</View>
`;
